[pkg/collector] Modularize `pkg/collector/runner/` logic

sgnn7 opened the pull request:
### What does this PR do?

This code should effectively contain almost no functional changes other
than splitting up the `Runner` object into components that can be easier
to understand, expand, and test. While some code has been improved and
cleaned up, the overall functionality of the previously-used APIs should
remain constant.

### Motivation

We need to add utilization expvars to each runner soon and the current architecture
seems to have reached its limits. This change should ensure that we can add this
functionality soon with fewer problems, fewer edge cases, more tests, and lower
maintenance costs.

### Additional Notes

- Before: `coverage: 54.8% of statements` (6 tests with single logic-path test so even the percentage here is misleading).
- After: `coverage: 96.4% of statements`

### Describe how to test your changes

1. Worker counts
   - Run the agent with a set number of workers (`check_runners` in config) and ensure that the agent starts with that many (grep for `ensureMinWorkers`).
   - Run the agent with no worker count specified (`check_runners` in config) and ensure that the agent starts with tthe default of 4 (grep for `ensureMinWorkers`).
1. Functionality
   - Ensure that all checks enabled report data to DD (e.g. `cpu` check via `system.cpu.user` metric)
1. Expvars / telemetry
   -  Get the expvars `curl -s localhost:5000/debug/vars | jq '.runner'` and ensure that the data fields report appropriate data:
      - Check that normal runs are recorded
      - Use a check that has errors and ensure that the errors are reported 
      - Use a check that has errors and ensure that the errors are reported in Agent log
   -  Get the telemetry (enable `telemetry.enabled: true` and `telemetry.checks: '*'`) then `curl -s localhost:5000/telemetry | grep ^checks` and ensure that the data fields report appropriate data:
      - Check that normal runs are recorded
      - Use a check that has errors and ensure that the errors are reported 
      - Use a check that has errors and ensure that the errors are reported in Agent log
1. Agent stop
   - Enable debug log and stop the agent with `<CTRL>-c`.
     - Ensure that runner is stopped (look for `Runner 1 shut down` in log)
     - Ensure that all workers are stopped (look for `Finished processing checks` in logs)
     - Ensure that the agent exits the process
   - Use a check that takes some time to run and `<CTRL>-c` during its run. Ensure that the agent exits gracefully.
1. Agent service checks
   - Ensure that we record the service checks (grep for `sendServiceChecks` in logs)
   - Ensure that on shutdown we record the service checks (grep for `sendServiceChecks` in logs)

### Checklist

- [ ] The appropriate `team/..` label has been applied, if known.
- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.

sgnn7 committed:
Modularize `pkg/collector/runner` and add tests

This code should effectively contain almost no functional changes other
than splitting up the `Runner` object into components that can be easier
to understand, expand, and test. Tests were added for almost all logic
that did not have it previously.

hush-hush commented on pkg/collector/runner/expvars.go:
```diff
@@ -0,0 +1,174 @@
+		return nil, false
+	}
+
+	return stats[id], true
```
Are we sure that `stats` will always contain `id`?

hush-hush commented on pkg/collector/runner/running_checks_tracker.go:
```diff
@@ -0,0 +1,75 @@
+
+// RunningChecksTracker is an object that keeps a thread-safe track of
+// all the running checks
+type RunningChecksTracker struct {
```
Do we need all those methods/Objects to be part of the public API of the `runner` package ?

sgnn7 replied:
Made a bunch of the public methods/objects private

hush-hush reviewed (commented):
Looks good overall. I didn't check all the public method but we should only export the one used outside.

djmitche reviewed (commented):
There's some interesting interactions with locking of the RunningChecksTracker here.  I like the refactoring, but I'd take it one step further and actually create sub-packages for Worker, RunningChecksTracker, and the stats.  And to @hush-hush's point about public interfaces, some attention to the API of each of those packages would help iron some of these things out, and indicate where docs are required for surprising or unusual API attributes (e.g., "you must hold the lock on this object while calling methods A and B but not for C and D").

sgnn7 commented on pkg/collector/runner/runner.go:
```diff
@@ -30,65 +23,45 @@ const (
+		isRunning:            1,
+		pendingCheckChan:     make(chan check.Check),
+		runningChecksTracker: NewRunningChecksTracker(),
+		staticWorkerCount:    numWorkers != 0,
```
I did `staticWorkerCount` -> `isStaticWorkerCount` if that works for you to completely disambiguate

sgnn7 commented on pkg/collector/runner/runner.go:
```diff
@@ -131,44 +120,47 @@ func (r *Runner) UpdateNumWorkers(numChecks int64) {
-		r.AddWorker()
-		numWorkers++
-		added++
+	if desiredNumWorkers <= int(atomic.LoadUint64(&r.workers)) {
```
This code previously depended on using values from expvars for `numWorkers` and used a static loop variable target for incrementing workers, making it thread-unsafe as well as multi-instance-unsafe. This change should fix both (if I do it well enough).

sgnn7 commented on pkg/collector/runner/expvars.go:
```diff
@@ -0,0 +1,174 @@
+
+// expvarRunnerCheckStats holds the stats from the running checks
+type expvarRunnerCheckStats struct {
+	Stats     map[string]map[check.ID]*check.Stats
```
Yeah that's fair - I was mostly focused on splitting things out, not changing the output format initially but that's a good suggestion

djmitche replied:
I think you could keep the output format the same, and only change how it's stored..

sgnn7 replied:
I deferred this change for now given the scope of changes already but let me know if you want me to get this implemented.

sgnn7 commented on pkg/collector/runner/running_checks_tracker.go:
```diff
@@ -0,0 +1,75 @@
+
+// RunningChecks returns a list of all the running checks
+func (t *RunningChecksTracker) RunningChecks() map[check.ID]check.Check {
+	// TODO: Copy the list?
```
Closures/anonymous functions aren't a really common pattern/standard for Golang in my experience but in this case it would make sense to do it that way. 👍

djmitche commented on pkg/collector/runner/tracker/tracker.go:
```diff
@@ -0,0 +1,106 @@
+)
+
+// withCheckFunc is a closure you can run on a mutex-locked check
+type withCheckFunc func(check.Check)
```
🥇  :)

djmitche commented on pkg/collector/runner/tracker/tracker.go:
```diff
@@ -0,0 +1,106 @@
+	runningChecksCopy := t.RunningChecks()
+
+	t.accessLock.Lock()
+	defer t.accessLock.Unlock()
```
Why copy _and_ lock?

djmitche reviewed (commented):
Looking great!

sgnn7 commented on pkg/collector/runner/expvars/timestamp_test.go:
```diff
@@ -0,0 +1,27 @@
+
+func TestTimestamp(t *testing.T) {
+	loc, err := time.LoadLocation("America/New_York")
+	if !assert.Nil(t, err) {
```
TIL. If this works, I think I'll be changing a bunch of my test code 😆

sgnn7 replied:
Fixed most of them (will push at EOD). I left a few that had improved error messages in some looping assertions.

djmitche commented on pkg/collector/runner/expvars/expvars.go:
```diff
@@ -0,0 +1,267 @@
+// Functions relating to check run stats (`expvarCheckStats`)
+
+func expCheckStatsFunc() interface{} {
+	return GetCheckStats
```
That says it calls the function.  It doesn't say that if the function returns another function, it will call that too!

sgnn7 replied:
Ah I see your point! Maybe I'll make sure this is explicit though - it's remnants from old code.

sgnn7 replied:
I wasn't able to really make this more elegant sadly so I'll leave it as-is

sgnn7 committed:
Fix test async data races

Seems like the `--race` had a few issues with the code that were all
legitimate so this commit fixes them.

sgnn7 committed:
[collector] Changed a few `sync.Mutex` objects to `sync.RWMutex`

Not everything needed a write lock for access so we now use a
more-optimized version of these in the refactored files.

sgnn7 committed:
[collector] Add tests for the scheduler usage

This code had no tests and this change adds them to exercise the logic.

sgnn7 committed:
[collector] Add more tests around `runner.Stop()`

With this change, we now have more tests around runner.Stop()
functionality that enhances our confidence that the code behaves as
expected.

sgnn7 committed:
[collector] Simplify `runner.Stop()` logic

Since we already had a `runner.StopCheck()` function, it didn't make
sense to have duplicate logic to complete all check stops so now we
ensure that we just call individual `StopCheck`s in the global `Stop`

sgnn7 committed:
[tests] Remove unnecessary instances of copying the index value in `tracker_test.go`

Since we only need to copy the values when the usage of a variable is
within a goroutine, there were some spots where the copy was
superfluous.

sgnn7 committed:
[collector] Fix `runner` expvar

This value was not correctly invoked (including tests) so this change
fixes it to match old behavior.

sgnn7 committed:
[collector] Fix flaky tests

We don't have a good notion of `<=` and `>=` for time so our check start
times now have padding to prevent flakyness.

sgnn7 committed:
[tests] Align test atomics to 64 bit boundaries

We had a test atomic that wa 32-bit aligned, breaking the tests so this
change ensures that the values are 64 bit and at the top of the struct
definition.

sgnn7 committed:
Merge remote-tracking branch 'origin/main' into sgnn7/runner-utilization-expvar-draft

olivielpeau commented on pkg/collector/runner/runner.go:
```diff
@@ -131,107 +168,112 @@ func (r *Runner) UpdateNumWorkers(numChecks int64) {
+
+		for _, c := range runningChecks {
+			// Save the check since the goroutine will reuse the pointer otherwise
+			ch := c
```
I might be missing some subtlety here, but instead of doing this, couldn't the closure that runs in a separate goroutine below take `c` a parameter? i.e.:

```
go func(ch check.Check) {
```

sgnn7 replied:
The approach I used is mostly muscle memory that I always used on these so that it's explicit about the loop variable gotcha. I'll see about changing this.

olivielpeau commented on pkg/collector/worker/check_logger.go:
```diff
@@ -0,0 +1,99 @@
+	log.Debugc("Running check...", "check", cl.Check)
+}
+
+// CheckFinshed is used to log that the check has completed
+func (cl *CheckLogger) CheckFinshed() {
```
typo:

```suggestion
// CheckFinished is used to log that the check has completed
func (cl *CheckLogger) CheckFinished() {
```

olivielpeau reviewed (commented):
I haven't reviewed in details, but at a high level this makes sense to me, thanks for the refactor and the tests! Just left a couple of comments. (no need for further approval from me)

sgnn7 committed:
[collector] Fix minor PR review issues

- Fix/remove goroutine saving of loop variables
- Fix spelling error in CheckFinished

djmitche commented on pkg/collector/runner/runner_test.go:
```diff
@@ -24,6 +24,10 @@ import (
 type testCheck struct {
+	// We have to align the atomics to 64-bit boundaries so these are put down first
+	runCount uint64
+	stopped  uint64
```
Per discussion, it looks like this is a [bug in staticcheck](https://staticcheck.io/docs/checks#SA1027), since (a) this wasn't a 32-bit access and (b) it was 64-bit aligned.. but fixing it by moving fields is easy.
